File: image/imagetoasm/imagetoasm.c
#include "imagetoasm.h"
#include <stdlib.h>
#include <getopt.h>
#include <string.h>

#define BITS 8
#define LABEL "label"
#define OFFSET 0
#define SKIP 1
#define SIZE 16
#define RAW 0
#define ASM 1
#define PGM 2
#define VERSION "1.01.01"
#define DATE "20190215"

void help(char *);
void version(void);

/*
 * imagetoasm: a tool for splitting image files into assembly code
 *
 * Theodore (Alex) Evans - 2019
 */

int main(int argc, char *argv[]) {
  struct option opts[] = {
    { "asm", 0, NULL, 'a' },
    { "bits", 1, NULL, 'b' },
    { "pgm", 0, NULL, 'g' },
    { "help", 0, NULL, 'h' },
    { "in", 1, NULL, 'i' },
    { "label", 1, NULL, 'l' },
    { "out", 1, NULL, 'o' },
    { "offset", 1, NULL, 'O' },
    { "pal", 1, NULL, 'p' },
    { "raw", 0, NULL, 'r' },
    { "sprite", 0, NULL, 's' },
    { "skip", 1, NULL, 'S' },
    { "tile", 0, NULL, 't' },
    { "version", 0, NULL, 'V' },
    { "verbose", 0, NULL, 'v' },
    { "size", 1, NULL, 'z' },
    { NULL, 0, NULL, 0 }
  };
  char optstring[]="ab:ghi:l:o:O:p:rsS:tvVz:";
  int opt, ix;
  FILE *infile, *outfile, *palfile;
  struct pam *inpam;
  int verbose=0;
  int offset=OFFSET;
  int skip=SKIP;
  int size=SIZE;
  int bits=BITS;
  int type=ASM;
  pal_t pal;
  rgb_t rgb, dec;
  ixed_t ixed;
  char *label=LABEL;

  infile=stdin;
  outfile=stdout;
  palfile=NULL;
  pnm_init(&argc, argv);
  inpam=(struct pam *)calloc(sizeof(struct pam), 1);
  while((opt=getopt_long(argc, argv, optstring, opts, &ix))!=-1) {
    switch(opt) {
    case 'b':
      bits=atoi(optarg);
      break;
    case 'g':
      type=PGM;
      break;
    case 'h':
      help(argv[0]);
      exit(0);
    case 'i':
      if(!(infile=fopen(optarg, "r"))) {
	fprintf(stderr, "Unable to open %s\n", optarg);
	return 1;
      }
      break;
    case 'l':
      label=strdup(optarg);
      break;
    case 'o':
      if(!(outfile=fopen(optarg, "w"))) {
	fprintf(stderr, "Unable to open %s\n", optarg);
	return 1;
      }
      break;
    case 'O':
      offset=atoi(optarg);
      break;
    case 'p':
      if (!(palfile=fopen(optarg, "r"))) {
	fprintf(stderr, "Unable to open %s\n", optarg);
	return 1;
      }
      break;
    case 'r':
      type=RAW;
      break;
    case 's':
      bits=8;
      size=16;
      break;
    case 'S':
      skip=atoi(optarg);
      break;
    case 't':
      bits=4;
      size=8;
      break;
    case 'V':
      version();
      exit(0);
    case 'v':
      verbose++;
      break;
    case 'z':
      size=atoi(optarg);
      break;
    default:
      fprintf(stderr, "Illegal argument %c\n", opt);
      help(argv[0]);
      exit(1);
    }
  }
  for(int i=0; i<argc; i++) fprintf(stderr, "%s ", argv[i]);
  putc('\n', stderr);
  if (optind<argc && infile==stdin) {
    if (!(infile=fopen(argv[optind], "r"))) {
      fprintf(stderr, "Unable to open %s\n", argv[optind]);
      return 1;
    }
    optind++;
  }
  if (optind<argc && outfile==stdout) {
    if (!(outfile=fopen(argv[optind], "w"))) {
      fprintf(stderr, "Unable to open %s\n", argv[optind]);
      return 1;
    }
    optind++;
  }
  set_verbose(verbose);
  pnm_readpaminit(infile, inpam, sizeof(struct pam));
  if (palfile) {
    pal=readpal(1<<bits, palfile);
    fclose(palfile);
  } else {
    pal=palette(bits);
  }
  if (!pal.l) return 1;
  rgb=readrgb(inpam);
  if (!rgb.dat) return 1;
  fclose(infile);
  dec=decimate(rgb, offset, offset, skip, skip);
  free_rgb(rgb);
  ixed=rgb2index(dec, pal);
  free_rgb(dec);
  free_pal(pal);
  switch (type) {
  case RAW:
    writeraw(outfile, ixed, size, size);
    break;
  case ASM:
    writeasm(outfile, ixed, size, size, label);
    break;
  case PGM:
    writepgm(outfile, ixed, size, size);
    break;
  default:
    fprintf(stderr, "Error\n");
  }
  free_ixed(ixed);
  fclose(outfile);
}

void help(char *name) {
  version();
  fprintf(stderr, "Usage: %s [<options>] [<infile>] [<outfile>]\n", name);
  fprintf(stderr, "\toptions with defaults in parenthesis:\n");
  fprintf(stderr, "\t-a\t--asm\t\toutput assembly code (asm)\n");
  fprintf(stderr, "\t-b\t--bits\t\tnumber of bits per pixel (%d)\n", BITS);
  fprintf(stderr, "\t-h\t--help\t\tshow this help message\n");
  fprintf(stderr, "\t-i\t--in\t\tinput file (stdin)\n");
  fprintf(stderr, "\t-l\t--label\t\tlabel in assembly file (%s)\n", LABEL);
  fprintf(stderr, "\t-o\t--out\t\toutput file (stdout)\n");
  fprintf(stderr, "\t-O\t--offset\toffset (x and y) in image (%d)\n", OFFSET);
  fprintf(stderr, "\t-p\t--pal\t\tpalette file (internal)\n");
  fprintf(stderr, "\t-r\t--raw\t\tout raw binary data (asm)\n");
  fprintf(stderr, "\t-s\t--sprite\tgenerate sprites = -b8 -z16\n");
  fprintf(stderr, "\t-S\t--skip\t\tuse ever nth pixel (%d)\n", SKIP);
  fprintf(stderr, "\t-t\t--tile\t\tgenerate tiles = -b4 -z8\n");
  fprintf(stderr, "\t-V\t--version\tshow version information\n");
  fprintf(stderr, "\t-v\t--verbose\tincrease verbosity\n");
  fprintf(stderr, "\t-z\t--size\t\telement size (%d)\n", SIZE);
}

void version(void) {
  fprintf(stderr, "imagetoasm version %s %s\n", VERSION, DATE);
  libzxntoolsver(1);
}

File: image/python/imagetoasm.py
#!/usr/bin/env python3
import getopt
import sys

from PIL import Image

import zxntools as zxn

BITS = 8
LABEL = "label"
OFFSET = 0
SKIP = 1
SIZE = 16
RAW = 0
ASM = 1
PGM = 2
VERSION = "1.00.00"
DATE = "20210324"


def my_help(name):
    version()
    sys.stderr.write(
        """Usage: {} [<options>] [<infile>] [<outfile>]
\toptions are
\t-a\t--asm\t\toutput assembly code (asm)
\t-b\t--bits\t\tset bit depth of palette ({})
\t-h\t--help\t\tshow this help message
\t-i\t--in\t\tinput file (stdin)
\t-l\t--label\t\tlabel in assembly file({})
\t-o\t--out\t\toutput file (stdout)
\t-O\t--offset\toffset (x and y) in image ({})
\t-p\t--pal\t\tpalette file (internal)
\t-r\t--raw\t\toutput raw binary data (asm)
\t-s\t--sprite\tgenerate sprites = -b8 -z16
\t-S\t--skip\t\tuse every nth pixel ({})
\t-V\t--version\tget version information
\t-v\t--verbose\tincrease verbosity
\t-z\t--size\t\telement size ({})
""".format(name, BITS, LABEL, OFFSET, SKIP, SIZE))


def version():
    sys.stderr.write("imagetoasm version {} {}\n".format(VERSION, DATE))
    zxn.version(True)


def decimate(img, offx, offy, skipx, skipy):
    inx, iny = img.size
    outx = (inx - offx + skipx - 1) // skipx
    outy = (iny - offy + skipy - 1) // skipy
    rv = Image.new('RGB', (outx, outy))

    y2 = offy
    for y1 in range(outy):
        x2 = offx
        for x1 in range(outx):
            rv.putpixel((x1, y1), img.getpixel((x2, y2)))
            x2 += skipx
        y2 += skipy

    return rv


def writeraw(fd, img, bits, txsz, tysz):
    ixsz, iysz = img.size
    tx = ixsz // txsz
    ty = iysz // tysz
    mask = (1 << bits) - 1
    buf = []
    for tr in range(ty):
        rbase = tr * tysz
        for tc in range(tx):
            cbase = tc * txsz
            for r in range(tysz):
                for c in range(0, txsz, 8 // bits):
                    v = 0
                    for b in range(8 // bits):
                        v = (v << bits) | img.getpixel((c + cbase + b, r + rbase)) & mask
                    buf.append(v)

    fd.write(bytes(buf))


def writepgm(fd, img, bits, txsz, tysz):
    ixsz, iysz = img.size
    tx = ixsz // txsz
    ty = iysz // tysz
    mask = (1 << bits) - 1
    fd.write(bytes("P5\n{} {}\n{}\n".format(txsz, ixsz * tx, mask), 'ascii'))
    buf = []
    for tr in range(ty):
        rbase = tr * tysz
        for tc in range(tx):
            cbase = tc * txsz
            for r in range(tysz):
                for c in range(0, txsz, 8 // bits):
                    for b in range(8 // bits):
                        buf.append(img.getpixel((c + cbase + b, r + rbase)) & mask)

    fd.write(bytes(buf))


def writeasm(fd, img, bits, label, txsz, tysz):
    ixsz, iysz = img.size
    tx = ixsz // txsz
    ty = iysz // tysz
    mask = (1 << bits) - 1

    fd.write("{}\n".format(label))
    cnt = 0
    col = 0

    for tr in range(ty):
        rbase = tr * tysz

        for tc in range(tx):
            cbase = tc * txsz
            if col != 0:
                fd.write("\n")
                col = 0
            fd.write(";; {} {:02x}\n".format(label, cnt))
            cnt += 1
            for r in range(tysz):
                for c in range(0, txsz, 8 // bits):
                    v = 0
                    for b in range(8 // bits):
                        v = (v << bits) | img.getpixel((c + cbase + b, r + rbase)) & mask
                    if col == 0:
                        fd.write("\tdefb {:02x}".format(v))
                    elif col != 15:
                        fd.write(", {:02x}".format(v))
                    else:
                        fd.write(", {:02x}\n".format(v))
                    col = (col + 1) % 16

    if col != 0:
        fd.write("\n")
    fd.write("{}_end:\n".format(label))


verbose = 0
offset = OFFSET
skip = SKIP
size = SIZE
bits = BITS
outtype = ASM
label = LABEL
inname = None
outname = None
palname = None
palfile = None

opts = "ab:ghi:l:o:O:p:rsS:tvVz:"
longopts = ['asm', 'bits=', 'pgm', 'help', 'in=', 'label=', 'out=', 'offset=',
            'pal=', 'raw', 'sprite', 'skip=', 'tile' 'version' 'verbose', 'size=']

try:
    optlist, arglist = getopt.getopt(sys.argv[1:], opts, longopts)
except getopt.GetoptError as err:
    sys.stderr.write("{}\n".format(err))
    exit(1)

for opt, arg in optlist:
    if opt in ('-b', '--bits'):
        bits = int(arg)
        if bits != 1 and bits != 2 and bits != 4 and bits != 8:
            sys.stderr.write("Unsupported bit depth: {}\n".format(bits))
            exit(1)

    elif opt in ('-g', '--pgm'):
        outtype = PGM

    elif opt in ('-h', '--help'):
        my_help(sys.argv[0])
        exit(0)

    elif opt in ('-i', '--in'):
        inname = arg

    elif opt in ('-l', '--label'):
        label = arg

    elif opt in ('-O', '--offset'):
        offset = int(arg)

    elif opt in ('-o', '--out'):
        outname = arg

    elif opt in ('-p', '--pal'):
        palname = arg

    elif opt in ('-r', '--raw'):
        outtype = RAW

    elif opt in ('-s', '--sprite'):
        bits = 8
        size = 16

    elif opt in ('-S', '--skip'):
        skip = int(arg)

    elif opt in ('-t', '--tile'):
        bits = 4
        size = 8

    elif opt in ('-V', '--version'):
        version()
        exit(0)

    elif opt in ('-v', '--verbose'):
        verbose += 1

    elif opt in ('-z', '--size'):
        size = int(arg)

    else:
        sys.stderr.write("Illegal option {}\n", opt)
        exit(1)

if palname is not None:
    try:
        palfile = open(palname, 'rb')
    except OSError:
        sys.stderr.write("Unable to open {}\n".format(inname))
        exit(1)

if len(arglist) > 0 and inname is None:
    inname = arglist.pop(0)

if len(arglist) > 0 and outname is None:
    outname = arglist.pop(0)

if inname is None:
    sys.stderr.write("No input file\n")

if outname is None:
    sys.stderr.write("No output file\n")

try:
    infile = open(inname, 'rb')
except OSError:
    sys.stderr.write("Unable to open {}\n".format(inname))
    exit(1)

try:
    if outtype == RAW or outtype == PGM:
        outfile = open(outname, 'wb')
    else:
        outfile = open(outname, 'w')
except OSError:
    sys.stderr.write("Unable to open {}\n".format(outname))
    exit(1)

zxn.setverbose(verbose)
image = Image.open(infile)
image.convert('RGB')
# rgb = image.data()

if palfile is not None:
    pal = zxn.readpal(palfile)
else:
    pal = zxn.makepal(zxn.palette(bits))

dec = decimate(image, offset, offset, skip, skip)
dec = dec.quantize(colors=(1 << bits), palette=pal, dither=Image.NONE)

xsz, ysz = dec.size
if xsz % size != 0 or ysz % size != 0:
    sys.stderr.write("Not perfectly tilable\n")
    exit(-1)

if outtype == RAW:
    writeraw(outfile, dec, bits, size, size)
elif outtype == ASM:
    writeasm(outfile, dec, bits, label, size, size)
elif outtype == PGM:
    writepgm(outfile, dec, bits, size, size)
else:
    sys.stderr.write("Output type error\n")

infile.close()
outfile.close()
